### PR TITLE
feat(@aws-amplify/ui-components): Trim inputs before submit

### LINKED ISSUE #6737: User friendly invalid username error
**Describe the bug**
I am still seeing the behaviour described by kennu in  [User friendly invalid username error #3862](https://github.com/aws-amplify/amplify-js/issues/3862)  when a user leaves a trailing space after their user name.
**To Reproduce**
Attempt to log in using Amplify UI with a trailing space after username.
**Desired behaviour**
Present the user with a more user friendly error message.

image

### PR DESCRIPTION
**Pull request by wlee221**
_Issue #, if available:_ Fixes #6737

_Description of changes:_ When user inputs have trailing or leading whitespace, Cognito returns 
```
Value at 'username' failed to satisfy constraint: Member must satisfy regular expression pattern: [\p{L}\p{M}\p{S}\p{N}\p{P}]+
```
which is not readable to end users. This pr trims inputs when forms are submitted to avoid these errors. We do not trim password and instead throw an error. 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-ui-components/src/common/helpers.ts [hunk 1]
```diff
@@ -15,8 +15,8 @@ import Auth from '@aws-amplify/auth';
 const logger = new Logger('helpers');
 
 export interface ToastError {
-  code: string;
-  name: string;
+  code?: string;
+  name?: string;
```
Comment (wlee221) [inline packages/amplify-ui-components/src/common/helpers.ts:19]: These fields should be optional, as we do not expect them to be present in our code.

**Comment by isaiahtaylorhh** [comment]
We ran into this issue last week. Very nasty error when it comes up! Thanks for the PR.

File: packages/amplify-ui-components/src/components/amplify-sign-up/amplify-sign-up.tsx [hunk 2]
```diff
@@ -131,7 +131,10 @@ export class AmplifySignUp {
       default:
         break;
     }
-
+    this.signUpAttributes.username = this.signUpAttributes.username.trim();
+    if (this.signUpAttributes.password !== this.signUpAttributes.password.trim()) {
+      dispatchToastHubEvent(new Error(Translations.PASSWORD_REMOVE_WHITESPACE));
+    }
```
Comment (wlee221) [inline packages/amplify-ui-components/src/components/amplify-sign-up/amplify-sign-up.tsx:137]: For passwords, we do not trim and instead throw an error

File: packages/amplify-ui-components/src/components/amplify-confirm-sign-up/amplify-confirm-sign-up.tsx [hunk 3]
```diff
@@ -214,6 +215,8 @@ export class AmplifyConfirmSignUp {
         break;
     }
     try {
+      if (!this.userInput) throw new Error('Username can not be empty');
```
Comment (sammartinez) [inline packages/amplify-ui-components/src/components/amplify-confirm-sign-up/amplify-confirm-sign-up.tsx:218]: Can we use the following? 
```suggestion
      if (!this.userInput) throw new Error(Translations.EMPTY_USERNAME);
```

**Review by sammartinez — COMMENTED** [review]
Overall, looks good. One callout to change and can we add a few unit tests please

File: packages/amplify-ui-components/src/components/amplify-confirm-sign-up/amplify-confirm-sign-up.tsx (hunk 3 above)
Comment (wlee221) [reply packages/amplify-ui-components/src/components/amplify-confirm-sign-up/amplify-confirm-sign-up.tsx:218]: Damn, can't believe I missed that one!

**Comment by wlee221** [comment]
@sammartinez, thanks for the review! I created test for this on the integration repo ([PR](https://github.com/aws-amplify/amplify-js-samples-staging/pull/188)).

**Comment by wlee221** [comment]
I also added stricter username checks in sign-up, as customers would be expecting their **exact** input to be used for sign up. For other components like sign-in, usernames are still trimmed for their convenience.

**Review by sammartinez — APPROVED** [review]
LGTM 🌮

**Pull request merged**